File: c++common/utils/Cache.cpp
#include "Cache.hpp"

Cache::Cache(std::chrono::seconds default_ttl)
    : default_ttl_(default_ttl) {}

void Cache::set(const std::string &key, const std::any &value, std::chrono::seconds ttl) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto expire = std::chrono::system_clock::now() + (ttl.count() == 0 ? default_ttl_ : ttl);
    cache_[key] = {value, expire};
}

std::optional<std::any> Cache::get(const std::string &key) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = cache_.find(key);
    if (it == cache_.end()) return std::nullopt;
    if (std::chrono::system_clock::now() > it->second.expire) {
        cache_.erase(it);
        return std::nullopt;
    }
    return it->second.value;
}

void Cache::remove(const std::string &key) {
    std::lock_guard<std::mutex> lock(mutex_);
    cache_.erase(key);
}

void Cache::clear() {
    std::lock_guard<std::mutex> lock(mutex_);
    cache_.clear();
}

std::size_t Cache::size() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return cache_.size();
}

File: Rust/common/utils/cache.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

pub struct Cache<T: Clone> {
    store: Mutex<HashMap<String, (T, Instant)>>,
    default_ttl: Duration,
}

impl<T: Clone> Cache<T> {
    pub fn new(default_ttl: Duration) -> Self {
        Self {
            store: Mutex::new(HashMap::new()),
            default_ttl,
        }
    }

    pub fn set(&self, key: &str, value: T, ttl: Option<Duration>) {
        let mut map = self.store.lock().unwrap();
        let expire = Instant::now() + ttl.unwrap_or(self.default_ttl);
        map.insert(key.to_string(), (value, expire));
    }

    pub fn get(&self, key: &str) -> Option<T> {
        let mut map = self.store.lock().unwrap();
        if let Some((val, exp)) = map.get(key) {
            if Instant::now() <= *exp {
                return Some(val.clone());
            } else {
                map.remove(key);
            }
        }
        None
    }

    pub fn delete(&self, key: &str) {
        self.store.lock().unwrap().remove(key);
    }

    pub fn clear(&self) {
        self.store.lock().unwrap().clear();
    }

    pub fn size(&self) -> usize {
        self.store.lock().unwrap().len()
    }
}
